### PR TITLE
ipsec: Fix key derivation regression due to boot IDs

### LINKED ISSUE #32040: gh: ipsec: support arbitrary number of nodes in ipsec-key-rotate action
In the per-tunnel key system, the `ipsec-key-rotate` action currently assumes a [fixed number](https://github.com/cilium/cilium/blob/a6b8538908f448a962e19730c6a34ece7c1caa81/.github/actions/ipsec-key-rotate/action.yaml#L46) of nodes. This is tuned to the `conformance-ipsec-e2e` workflow - but doesn't support the `conformance-eks` workflow, which uses a different number of nodes. Thus the eks workflow [currently doesn't test](https://github.com/cilium/cilium/blob/a6b8538908f448a962e19730c6a34ece7c1caa81/.github/workflows/conformance-eks.yaml#L321) the per-tunnel key system.

We should improve the key-rotate action, so that it supports workflows with any number of nodes.

### LINKED ISSUE #35693: CI: Conformance GKE fails on Wait for Cilium to be ready
CI: Conformance GKE fails constantly on Wait for Cilium to be ready. This issue might be related to or the same issue as these #34994 and #35508 

```
Error: Unable to determine status:  timeout while waiting for status to become successful: context deadline exceeded
    /¯¯\
 /¯¯\__/¯¯\    Cilium:             OK
 \__/¯¯\__/    Operator:           OK
 /¯¯\__/¯¯\    Envoy DaemonSet:    OK
 \__/¯¯\__/    Hubble Relay:       1 errors
    \__/       ClusterMesh:        disabled

DaemonSet              cilium             Desired: 2, Ready: 2/2, Available: 2/2
DaemonSet              cilium-envoy       Desired: 2, Ready: 2/2, Available: 2/2
Deployment             cilium-operator    Desired: 1, Ready: 1/1, Available: 1/1
Deployment             hubble-relay       Desired: 1, Unavailable: 1/1
Containers:            cilium             Running: 2
                       cilium-envoy       Running: 2
                       cilium-operator    Running: 1
                       hubble-relay       Running: 1
Cluster Pods:          0/14 managed by Cilium
Helm chart version:    1.17.0-dev
Image versions         cilium             quay.io/cilium/cilium-ci:e956866f5fef16192836e6280c9ba6529230bc02: 2
                       cilium-envoy       quay.io/cilium/cilium-envoy:v1.30.7-1730239547-90b7953f5b6ffe617fa630d42ea6d5b65fba1b9e@sha256:d7db101d856c478a497fc79fab3ce3c732221e80fdf37bfe001672acb155d164: 2
                       cilium-operator    quay.io/cilium/operator-generic-ci:e956866f5fef16192836e6280c9ba6529230bc02: 1
                       hubble-relay       quay.io/cilium/hubble-relay-ci:e956866f5fef16192836e6280c9ba6529230bc02: 1
Errors:                hubble-relay       hubble-relay    1 pods of Deployment hubble-relay are not ready
```

Example runs:
https://github.com/cilium/cilium/actions/runs/11609904397/job/32328150563
https://github.com/cilium/cilium/actions/runs/11627714241/job/32381575897

Artifacts:
[cilium-junits.zip](https://github.com/user-attachments/files/17599913/cilium-junits.zip)
[cilium-sysdumps.zip](https://github.com/user-attachments/files/17599912/cilium-sysdumps.zip)

### PR DESCRIPTION
**Pull request by pchaigno**
See commit descriptions. This bug made it only into a pre-release for v1.17.

Fixes: #35210.
```release-note
Fix bug that would break all pod-to-pod connectivity when using the per-tunnel IPsec key system.
```

### CODE REVIEW

File: pkg/datapath/linux/ipsec/ipsec_linux.go [hunk 1]
```diff
@@ -538,7 +538,7 @@ func xfrmKeyEqual(s1, s2 *netlink.XfrmState) bool {
 }
 
 func ipSecReplaceStateIn(log *slog.Logger, params *IPSecParameters) (uint8, error) {
-	key := getNodeIPsecKey(*params.SourceTunnelIP, *params.DestTunnelIP, params.LocalBootID, params.RemoteBootID)
+	key := getNodeIPsecKey(*params.SourceTunnelIP, *params.DestTunnelIP, params.RemoteBootID, params.LocalBootID)
```
Comment (tommyp1ckles) [inline pkg/datapath/linux/ipsec/ipsec_linux.go:541]: non-blocking-comment/re:internal discussion: the key derivation does seem easy to break. We probably should be able to catch regressions via unit tests and not relying on various e2e.

File: .github/actions/ipsec-key-rotate/action.yaml [hunk 2]
```diff
@@ -70,13 +62,9 @@ runs:
...
-        exp_nb_keys=1
-        if [[ "${{ inputs.key-type-two }}" == "+" ]]; then
-          exp_nb_keys=8
-          if [[ "${{ inputs.encryption-overlay }}" == "true" ]]; then
-            exp_nb_keys=12
-          fi
-        fi
+        # After the key rotation is finished and old keys are cleaned up, the number of keys is
+        # halved.
+        ((exp_nb_keys/=2))
```
Comment (tommyp1ckles) [inline .github/actions/ipsec-key-rotate/action.yaml:67]: How come this wasn't failing before?

**Review by tommyp1ckles — APPROVED** [review]

File: .github/actions/ipsec-key-rotate/action.yaml (hunk 2 above)
Comment (pchaigno) [reply .github/actions/ipsec-key-rotate/action.yaml:67]: If you look at the numbers, they were also halved before. It just was hardcoded.

File: .github/workflows/conformance-ipsec-e2e.yaml [hunk 3]
```diff
@@ -304,9 +304,9 @@ jobs:
         uses: ./.github/actions/ipsec-key-rotate
         with:
           key-algo: ${{ matrix.key-two }}
-          key-type-one: ${{ matrix.key-type-one }}
-          key-type-two: ${{ matrix.key-type-two }}
           encryption-overlay: ${{ matrix.encryption-overlay }}
+          nb-nodes: 2
+          ipv6: true
```
Comment (smagnani96) [inline .github/workflows/conformance-ipsec-e2e.yaml:309]: It took me a while to understand that a parameter not in matrix is as equal as if it was set to true and that this is actually the expected behavior for this test :+1: 

While [setting up Kind](https://github.com/cilium/cilium/pull/35806/files#diff-b77d9c4420c99e9bbac5afc52278677e846739ed4b7b4f7c1efdb24f1b69a249L189) we do [use](https://github.com/cilium/cilium/blob/6a9766b37af22f6b902e4acbf8a9f137f545b35b/.github/workflows/conformance-ipsec-e2e.yaml#L172
) `$matrix.ipv6`, but we don't have it set in [.github/actions/ipsec/configs.yaml](https://github.com/cilium/cilium/blob/6a9766b37af22f6b902e4acbf8a9f137f545b35b/.github/actions/ipsec/configs.yaml), hence resulting in `ipv6=""`.

[Later](https://github.com/cilium/cilium/actions/runs/11704540713/job/32597494835#step:9:2) in In `/ci-ipsec-e2e`:
```bash
IP_FAM="dual"
if [ "" == "false" ]; then
  IP_FAM="ipv4"
fi
echo params="--xdp --secondary-network \"\" [3](https://github.com/cilium/cilium/actions/runs/11704540713/job/32597494835#step:9:3) \"\" \"\" iptables $IP_FAM" >> $GITHUB_OUTPUT
```
resulting in ipv6 being used and that is why `ipv6: true`, right?

**Review by smagnani96 — APPROVED** [review]
Changes looks good, took me a bit to figure out CI -- my bad (I left a comment explaining where I was stuck).

File: .github/workflows/conformance-ipsec-e2e.yaml (hunk 3 above)
Comment (pchaigno) [reply .github/workflows/conformance-ipsec-e2e.yaml:309]: Yep, that's correct. The "IPv6 enabled by default" also puzzled me a bit, but I guess it matches what the agent does as well.

File: pkg/datapath/linux/ipsec/ipsec_linux.go (hunk 1 above)
Comment (pchaigno) [reply pkg/datapath/linux/ipsec/ipsec_linux.go:541]: Yup, that might be doable. I've opened https://github.com/cilium/cilium/issues/35843 for it.

**Review by nbusseneau — APPROVED** [review]

**Pull request merged**